v-next/hardhat/src/internal/builtin-plugins/network-manager/json-rpc-request-modifiers/json-rpc-request-modifier.ts

```diff
@@ -0,0 +1,136 @@
+import { FixedGasPrice } from "./gas-properties/fixed-gas-price.js";
+import { FixedGas } from "./gas-properties/fixed-gas.js";
+import { isResolvedHttpNetworkConfig } from "./utils.js";
+
+export class JsonRpcRequestModifier {
+  readonly #provider: EthereumProvider;
+  readonly #networkConfig: NetworkConfig;
+
+  // chainId
+  #chainIdValidator: ChainIdValidator | undefined;
+
+  // gas
+  #automaticGas: AutomaticGas | undefined;
+  #fixedGas: FixedGas | undefined;
+
+  // gas price
+  #automaticGasPrice: AutomaticGasPrice | undefined;
+  #fixedGasPrice: FixedGasPrice | undefined;
+
+  constructor(nextNetworkConnection: NetworkConnection<ChainType | string>) {
+    this.#provider = nextNetworkConnection.provider;
+    this.#networkConfig = nextNetworkConnection.networkConfig;
+  }
+
+  public async createModifiedJsonRpcRequest(
+    jsonRpcRequest: JsonRpcRequest,
+  ): Promise<JsonRpcRequest> {
+    // We clone the request to avoid interfering with other hook handlers that
+    // might be using the original request.
+    // The "newJsonRpcRequest" inside this class is modified by reference.
+    const newJsonRpcRequest = await deepClone(jsonRpcRequest);
+
+    await this.#modifyGasAndGasPriceIfNeeded(newJsonRpcRequest);
+
+    await this.#validateChainIdIfNeeded(newJsonRpcRequest);
+
+    return newJsonRpcRequest;
+  }
+
+  async #modifyGasAndGasPriceIfNeeded(
+    jsonRpcRequest: JsonRpcRequest,
+  ): Promise<void> {
+    if (
+      this.#networkConfig.gas === undefined ||
+      this.#networkConfig.gas === "auto"
+    ) {
+      if (this.#automaticGas === undefined) {
+        this.#automaticGas = new AutomaticGas(
+          this.#provider,
+          this.#networkConfig.gasMultiplier,
+        );
+      }
+
+      await this.#automaticGas.modifyRequest(jsonRpcRequest);
+    } else {
+      if (this.#fixedGas === undefined) {
+        this.#fixedGas = new FixedGas(this.#networkConfig.gas);
+      }
+
+      this.#fixedGas.modifyRequest(jsonRpcRequest);
+    }
+
+    if (
+      this.#networkConfig.gasPrice === undefined ||
+      this.#networkConfig.gasPrice === "auto"
+    ) {
+      // If you use a LocalAccountsProvider or HDWalletProvider, your transactions
+      // are signed locally. This requires having all of their fields available,
+      // including the gasPrice / maxFeePerGas & maxPriorityFeePerGas.
+      //
+      // We never use those providers when using Hardhat Network, but sign within
+      // Hardhat Network itself. This means that we don't need to provide all the
+      // fields, as the missing ones will be resolved there.
+      //
+      // Hardhat Network handles this in a more performant way, so we don't use
+      // the AutomaticGasPrice for it unless there are provider extenders.
+      // The reason for this is that some extenders (like hardhat-ledger's) might
+      // do the signing themselves, and that needs the gas price to be set.
+      if (isResolvedHttpNetworkConfig(this.#networkConfig)) {
```

Comment:
I need a double-check on this because, based on the V2 comment description above, I'm not entirely sure whether this "if condition" is correct. I'll update the V2 description as soon as I confirm the correct code.